### ISSUE TITLE
Do you have plan to implement unsupervised versions of GraphSAGE?

### ISSUE THREAD

**Issue by chenwgen** [williamleif/graphsage-simple#1]
Hi, do you have plan to implement unsupervised versions of GraphSAGE?  thanks.

**Comment by unsuthee** [comment]
You just need to change the loss function to Eq. 1 from the original paper.

**Comment by chithangduong** [comment]
@unsuthee have you implemented it? I have changed the loss function to Eq. 1 but the embeddings do not make sense: connected nodes do not have close embeddings.  They are different from the tensorflow version.

**Comment by HongxuChenUQ** [comment]
I've implemented the unsupervised version, by training the model using random walks or network edges. However, the converging is wired. Discussions are welcome.

**Comment by williamleif** [comment]
One thing that is necessary is to constrain the embeddings to be unit length. This is mentioned in the appendix of the paper I think. For instance you can use cosine instead of the dot product to achieve this. Though this is a minor thing, it can have a big impact on convergence.

**Comment by williamleif** [comment]
Sadly, I don't plan on implementing the unsupervised any time soon, but pull requests are welcome! :)

**Comment by HongxuChenUQ** [comment]
@williamleif That makes sense now!  Thanks for pointing out!

**Comment by fs302** [comment]
Hi @HongxuChenUQ, is it possible for you to share the loss code you made for unsupervised version?

Actually, I tried to combine loss_label and loss_network and found the F1 score lifts from 0.88 to 0.93. But when I leave the loss_network alone, **there will be none grad to the model's weight**. Since I am new to PyTorch, it is really annoying! I can't figure out the problem.

Below is my loss code, where nodes and negtive_samples are node lists.

```
def loss(self, nodes, negtive_samples, num_neighs, labels):
        loss_list = []
        z_negtive_samples = self.enc(negtive_samples).t()
        z_querys = self.enc(nodes).t()
        for i,query in enumerate(nodes):
            z_query = z_querys[i]
            neighbors = list(self.adj_lists[int(query)])[:num_neighs]
            z_neighbors = self.enc(neighbors).t()
            pos = torch.min(torch.sigmoid(torch.tensor([torch.dot(z_query,z_neighbor) for z_neighbor in z_neighbors]))).requires_grad_()
            neg = torch.max(torch.sigmoid(torch.tensor([torch.dot(z_query,z_ns) for z_ns in z_negtive_samples]))).requires_grad_()
            loss_list.append(torch.max(Variable(torch.tensor(0.0)),neg-pos+self.margin))
        loss_net = Variable(torch.mean(torch.tensor(loss_list)),requires_grad=True)
        scores = self.forward(nodes)
        loss_sup = self.xent(scores, labels.squeeze())
        return loss_sup+loss_net
```

**Comment by HongxuChenUQ** [comment]
`class UnsupervisedGraphSage(nn.Module):
    def __init__(self, enc):
        super(UnsupervisedGraphSage, self).__init__()
        self.enc = enc
        self.weight = nn.Parameter(torch.FloatTensor(1, enc.embed_dim)) # size 1*128 used for regression
        self.xent = nn.LogSigmoid()
        self.MSE = nn.MSELoss()
        self.sig = nn.Sigmoid()
        nn.init.xavier_uniform_(self.weight)

    def forward(self,u,v):
        embed_u = self.enc(u) #128*1024
        embed_v = self.enc(v) #128*1024
        #scores = embed_u.t().mm(embed_v).diag()
        scores = nn.functional.cosine_similarity(embed_u.t(),embed_v.t()) # Though this is a minor thing, it can have a big impact on convergence.
        return scores
    def pos_loss(self,u,v):
        scores = self.forward(u,v)
        return self.xent(scores)


    def neg_loss(self,nodes_u,neg_samples):
        neg_loss=0
        for node in nodes_u:
            embed_node = self.enc([node])#128*1
            embed_negs = self.enc(list(neg_samples[node]))#128 * negs
            scores = embed_node.t().mm(embed_negs)
            neg_loss += sum(self.xent(-scores).t())
        return neg_loss

    def loss(self,nodes_u,nodes_v,neg_samples):
        return -sum(self.pos_loss(nodes_u,nodes_v)) - self.neg_loss(nodes_u,neg_samples)
`

**Comment by HongxuChenUQ** [comment]
`
class UnsupervisedGraphSage(nn.Module):
    def __init__(self, enc):
        super(UnsupervisedGraphSage, self).__init__()
        self.enc = enc
        self.weight = nn.Parameter(torch.FloatTensor(1, enc.embed_dim)) # size 1*128 used for regression
        self.xent = nn.LogSigmoid()
        self.sig = nn.Sigmoid()
        nn.init.xavier_uniform_(self.weight)

    def forward(self,u,v):
        embed_u = self.enc(u) #128*1024
        embed_v = self.enc(v) #128*1024
        #scores = embed_u.t().mm(embed_v).diag()
        scores = nn.functional.cosine_similarity(embed_u.t(),embed_v.t()) # Though this is a minor thing, it can have a big impact on convergence.
        return scores

    def pos_loss(self,u,v):
        scores = self.forward(u,v)
        return self.xent(scores)


    def neg_loss(self,nodes_u,neg_samples):
        neg_loss=0
        for node in nodes_u:
            embed_node = self.enc([node])#128*1
            embed_negs = self.enc(list(neg_samples[node]))#128 * negs
            scores = embed_node.t().mm(embed_negs)
            neg_loss += sum(self.xent(-scores).t())
        return neg_loss

    def loss(self,nodes_u,nodes_v,neg_samples):
        return -sum(self.pos_loss(nodes_u,nodes_v)) - self.neg_loss(nodes_u,neg_samples)
`

**Comment by HongxuChenUQ** [comment]
hi @fs302: The following code is my class - UnsupervisedGraphSage(nn.Module):
`

    def __init__(self,enc):
        super(UnsupervisedGraphSage, self).__init__()
        self.enc = enc
        self.weight = nn.Parameter(torch.FloatTensor(1, enc.embed_dim)) # size 1*128 used for regression
        self.xent = nn.LogSigmoid()
        self.sig = nn.Sigmoid()
        nn.init.xavier_uniform_(self.weight)

    def forward(self,u,v):
        embed_u = self.enc(u) #128*1024
        embed_v = self.enc(v) #128*1024
        #scores = embed_u.t().mm(embed_v).diag()
        scores = nn.functional.cosine_similarity(embed_u.t(),embed_v.t()) # Though this is a minor thing, it can have a big impact on convergence.
        return scores

    def pos_loss(self,u,v):
        scores = self.forward(u,v)
        return self.xent(scores)


    def neg_loss(self,nodes_u,neg_samples):
        neg_loss=0
        for node in nodes_u:
            embed_node = self.enc([node])#128*1
            embed_negs = self.enc(list(neg_samples[node]))#128 * negs
            scores = embed_node.t().mm(embed_negs)
            neg_loss += sum(self.xent(-scores).t())
        return neg_loss

    def loss(self,nodes_u,nodes_v,neg_samples):
        return -sum(self.pos_loss(nodes_u,nodes_v)) - self.neg_loss(nodes_u,neg_samples)`

**Comment by fs302** [comment]
@HongxuChenUQ Really appreciate that!  What about your performance of F1 score?

**Comment by HongxuChenUQ** [comment]
@fs302 I've test it on AUC performance, it is good. You will have to train a classifier if you want to test it on F1 score.

**Comment by HongxuChenUQ** [comment]
@fs302 I've tested it on AUC performance, it is good. You will have to train a classifier if you want to test it on F1 score.

**Comment by fs302** [comment]
@HongxuChenUQ Yes, I use a 2-layer NN as downstream classifier, but only achieve F1=0.31, which is much lower than End-to-End supervised version(F1=0.84) on the same embedding setting. 

I wonder if it might be the difference of positive & negative pair sampling.

**Comment by fs302** [comment]
Thanks to @HongxuChenUQ, after tuning the learning rate of downstream classifier and generate a robust negative samples, the best F1 hit 0.76 for unsupervised version.

**Comment by tonygrey** [comment]
@HongxuChenUQ  Can you please explain the need  self.base_model = base_model in the encoders?

**Comment by ghost** [comment]
@fs302 @ @HongxuChenUQ  is it possible to share the code for unsupervised version?
